### ISSUE TITLE
[HOLD for payment 2024-10-17] [$250] Next Steps - "Waiting for userAdmin to finish setting up a business bank account" is not displayed with the correct conditions

### ISSUE THREAD

**Issue by isagoico** [Expensify/App#47271]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://github.com/Expensify/Expensify/issues/417883 - Being added here

**Email or phone of affected tester (no customers):** [applausetester+ig8@applause.expensifail.com](mailto:applausetester+ig8@applause.expensifail.com)

**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Issue reported by:** Applause - Internal team

## Action Performed:
_Preconditions:_ 
- Workspace approval & payments are enabled
- Bank account setup has a "Unfinished" setup and bank account is pending validation.
- Employee has submitted a report > the report is approved and is "awaiting payment"

1. Navigate to the workspace chat
1. Click on the report that is awaiting payment
2. Check the next steps banner

## Expected Result:
The next steps banner indicates "Waiting for userAdmin to finish setting up a business bank account" when a report is pending payment and the admin has not finished setting up the business bank account in the workspace. 

## Actual Result:
"Waiting for [userAdmin] to pay expense(s)." is displayed. 

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

![image](https://github.com/user-attachments/assets/b773135d-1ba4-4478-9173-9628f17c6b1b)

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~018c20eb05dc94ed77</li>
        <li>Upwork Job ID: 1823304297972705293</li>
        <li>Last Price Increase: 2024-08-13</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @dylanexpensify</details>

**Comment by dylanexpensify** [comment]
Reviewing today!

**Comment by c3024** [comment]
>We can use the check reimbursementAccount?.achData?.state === 'SETUP'

When the account adding process has not been started, there is no field of `achData` in `reimbursementAccount`. Could you clarify all the conditions in which we will show this message of `Waiting for userAdmin to finish setting up a business bank account`.

**Comment by dangrous** [comment]
Is this issue just in the optimistic (front end) next step? Or is the one coming from the server also incorrect?

**Comment by dylanexpensify** [comment]
@c3024 can you confirm? cc @isagoico

**Comment by c3024** [comment]
It seems that we are adding the message that appears here only optimistically.

When a user submits a request or an approver approves it, the message in `nextStep` is hardcoded until the next actor (approver or payer) acts (either approves or pays).

For example, if we change the `nextStep` message to `Waiting for admin to finish setting up a bank account` for cases where an account setup is pending, the following sequence would occur:

1. A user submits a request.
2. The approver approves the expense.
3. The admin payer sees `Waiting for admin to finish setting up a bank account`.
4. The admin sets up their bank account.

Even after the bank account is set up, the message doesn't update because the message in `nextStep` only changes when the next user (in this case, the admin) takes action (e.g., paying the expense).

To address this, we may need to change the logic for displaying the message so that it depends solely on the account setup state if we want to show a different message when the account is not set up.

However, in my opinion, it might not be worth the effort to fix this bug. It may be better to leave the message as `Waiting for the admin to pay expenses` and not make any changes.

**Comment by dangrous** [comment]
So does the backend sending out `waiting for admin to finish setting up a bank account` not update it for the report? I don't think we need to handle it optimistically necessarily, but if it's also not updating when the backend does, we should look into it.

**Comment by dangrous** [comment]
Sounds like we're _probably_ going to close this out as won't fix but confirming a couple things internally

**Comment by dylanexpensify** [comment]
@dangrous we're going to keep open based on this internal [thread](https://expensify.slack.com/archives/C06ML6X0W9L/p1723837995358059?thread_ts=1723491174.174309&cid=C06ML6X0W9L). Can we prioritize getting this sussed out so we can get the regression tests posted and wrap up sent? 🙏

**Comment by dangrous** [comment]
oh whoops I got confused. I can update the backend to send "Waiting for x to pay expenses." To clarify, should we not be sending that in any situation? Or just in this particular one?

**Comment by dangrous** [comment]
PR is up for back end, still confirming whether or not it's the best fix in [that Slack thread](https://expensify.slack.com/archives/C03TQ48KC/p1724256605429969?thread_ts=1724100808.197469&cid=C03TQ48KC) but hopefully should be in review soon

**Comment by dangrous** [comment]
backend is merged, not yet deployed. Once deployed, @c3024, are you interested in figuring out that front end logic? Or would you prefer we wait for contributor proposals

**Comment by dangrous** [comment]
Not overdue, waiting for BE PR to hit staging.

**Comment by dangrous** [comment]
Okay backend PR is up on staging. @c3024 are you interested in knocking out the front end logic for this? Or should we fish for a contributor in Slack/elsewhere?

**Comment by c3024** [comment]
I can start working on this tomorrow.

**Comment by dylanexpensify** [comment]
Nice, thanks @c3024!!

**Comment by c3024** [comment]
In which cases does the backend send the `nextStep` message as "Waiting for setting up a bank account"?

<img width="440" alt="Screenshot 2024-08-31 at 6 09 21 PM" src="https://github.com/user-attachments/assets/8d00edc2-50b7-489c-b6a0-e703ab9f0dde">

This is what I am seeing when connected to the staging server. The `state` in `achData` of `reimbursementAccount` shows as `SETUP`, but the `nextStep` message has the messages array as "Waiting to pay expense."

I think when this value is `SETUP`, the message in the next step should be "Waiting for setting up a bank account."

**Comment by dangrous** [comment]
So the backend will send the setup message if the bank account doesn't exist, or if the state is anything except `OPEN` - but it's looking at the `preferredReimbursementAccount` on the policy, which may be the difference here. Is that data the front end has access to?

We did it that way since it seemed to be the most consistent in showing the right message - but maybe there are some cases we missed.

**Comment by c3024** [comment]
This field `preferredReimbursementAccount` does not seem to be present in the `App` repo.

I'm following the details mentioned [here](https://github.com/Expensify/App/issues/36512#issuecomment-1954959865) to connect a bank account to the workspace. However, if I enter incorrect details and navigate to the Onfido page, only then does the `state` in `achData` show as `SETUP`. When no account is added, or when the bank account setup is at an intermediate stage, such as the profile page, there is no `achData` in the `policy`.

I'm unsure how to proceed from this point.

**Comment by dangrous** [comment]
Hm, well I can at least have the backend also check if `achData` `state` is `SETUP`... so we can fix that part.

What gets sent from the backend (or stored from the front end) between the bank account setup steps? There must be something we can look for, right?

**Comment by c3024** [comment]
When the bank account is fully set up, `reimbursementAccount?.achData?.state` is `open`. 

For all other states, whether the policy owner has not started setting up the bank account or is at an intermediate step, can we display "Waiting for setting up of bank account?"

**Comment by dangrous** [comment]
That makes sense to me! @dylanexpensify can you think of any weird edge cases where that wouldn't be ideal?

**Comment by dangrous** [comment]
bump @dylanexpensify on https://github.com/Expensify/App/issues/47271#issuecomment-2341454487 - I'm pretty sure that should be fine!

**Comment by dangrous** [comment]
@c3024 let's go ahead with that for now and we can always adjust later on - thank you!

**Comment by c3024** [comment]
Thanks. I will start working on the PR.

**Comment by dangrous** [comment]
How are we looking on this one? @alitoshmatov I think you got assigned to review so I'm adding you on this issue here. Thanks!

**Comment by dangrous** [comment]
Actually @alitoshmatov can you comment so I can assign you?

**Comment by c3024** [comment]
@DylanDylann [will review](https://expensify.slack.com/archives/C02NK2DQWUX/p1727873547784629) the PR.

**Comment by dangrous** [comment]
Assigned @DylanDylann to the PR for review and here as well!

**Comment by DylanDylann** [comment]
Sure

**Comment by dylanexpensify** [comment]
Mind giving us an update @DylanDylann 🙇‍♂️

**Comment by DylanDylann** [comment]
Waiting for @dangrous to give the final review on PR

**Comment by dylanexpensify** [comment]
Tomorrow!

**Comment by DylanDylann** [comment]
BugZero Checklist: The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

 [@DylanDylann] The PR that introduced the bug has been identified. Link to the PR: **New requirement, not a bug**
 [@DylanDylann] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: **NA**
 [@DylanDylann] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: **NA**
 [@DylanDylann] Determine if we should create a regression test for this bug. **Yes**
 [@DylanDylann] If we decide to create a [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) for the bug, please propose the regression test steps to ensure the same bug will not reach production again.


Regression Test Proposal

1. [User A] Create a policy with approvals enabled and approver as User A. Do not connect any account to this policy.
2. [User B] Request an expense in the workspace chat.
3. [User A] Approve the expense.
4. [User A] Go to the expense report, if not opened already.
5. [User A] Verify that the message below the header says "Waiting for to finish setting up a bank account".

Do we agree 👍 or 👎

**Comment by dylanexpensify** [comment]
@c3024 @DylanDylann sent offers - please accept!

**Comment by c3024** [comment]
Accepted, thanks!

**Comment by dylanexpensify** [comment]
@DylanDylann please accept!

**Comment by dylanexpensify** [comment]
Hey @adelekennedy! I'm heading out on parental leave so reassigning this! All that's left is paying out @DylanDylann once they accept their offer. Job post found [here](https://www.upwork.com/ab/applicants/1846981405054955746/hired)! TY! 🙇‍♂️

**Comment by adelekennedy** [comment]
Payment made!
